ansible_runner/receptor_plugin.py

```diff
@@ -0,0 +1,164 @@
+
+    async def read_responses():
+        while True:
+            message = await controller.recv()
+            if message.header.get("eof", False):
+                break
+            elif message.payload:
+                content = json.loads(message.payload.readall())
+                c_header = content[0]
+                c_type = c_header['type']
+                if c_type == 'event':
+                    data = content[1]
+                    if 'event_handler' in run_options and run_options['event_handler']:
+                        run_options['event_handler'](data)
+                    if 'stdout' in data and data['stdout']:
+                        print(data['stdout'])
+                elif c_type == 'status':
+                    data = content[1]
+                    if 'status_handler' in run_options and run_options['status_handler']:
+                        run_options['status_handler'](data, None)
+                elif c_type == 'error':
+                    print("Error from remote:", content[1])
+
+    async def run_func():
+        if receptor_node_id != receptor_node:
+            add_peer_task = controller.add_peer(receptor_peer)
+            start_wait = time.time()
+            while True:
+                if add_peer_task and add_peer_task.done() and not add_peer_task.result():
+                    return False
+                if controller.receptor.router.node_is_known(receptor_node):
+                    break
+                if time.time() - start_wait > 5:
+                    if not add_peer_task.done():
+                        add_peer_task.cancel()
+                    return False
+                await asyncio.sleep(0.1)
+        await controller.send(payload=tmpf.name, recipient=receptor_node, directive='ansible_runner:execute')
+        await controller.loop.create_task(read_responses())
+
+    remote_options = {key: value for key, value in run_options.items() if key in (
+        'forks',
+        'host_pattern',
+        'ident',
+        'ignore_logging',
+        'inventory',
+        'limit',
+        'module',
+        'module_args',
+        'omit_event_data',
+        'only_failed_event_data',
+        'playbook',
+        'verbosity',
+    )}
+
+    with tempfile.NamedTemporaryFile(suffix='.tgz') as tmpf:
+
+        # Create tar file
+        with tarfile.open(fileobj=tmpf, mode='w:gz') as tar:
+            if 'private_data_dir' in run_options:
+                tar.add(run_options['private_data_dir'], arcname='')
+            kwargs = json.dumps(remote_options, cls=UUIDEncoder)
+            ti = tarfile.TarInfo('kwargs')
+            ti.size = len(kwargs)
+            ti.mtime = time.time()
+            tar.addfile(ti, io.BytesIO(kwargs.encode('utf-8')))
+        tmpf.flush()
+
+        # Run the job via Receptor
+        if receptor_node_id:
+            receptor_args = f"--node-id {receptor_node_id} node --server-disable".split()
```

Comment:
*Is* runner at this point dedicated to py3-only?  I realize that if you're using this feature, and thus receptor, you're sort of locked into python3, but I wonder if this f-string is going to get get flagged as a syntax error for those using runner under py2?

Reply:
Runner is still being linted on python 2.7.  If it doesn't need to be, then that would surely make things easier.  As far as the f-string, on python 2.7, this module doesn't get imported at all.  See https://github.com/ansible/ansible-runner/blob/6c37d73c1ddbf34c829bd3adf8280ee1c7a12054/ansible_runner/__main__.py#L45.